numpy/core/src/npymath/npy_math_complex.c.src

```diff
@@ -445,29 +445,48 @@ npy_cpow@c@ (@ctype@ a, @ctype@ b)
     @type@ bi = npy_cimag@c@(b);
     @ctype@ r;
 
+    /*
+     * Checking if in a^b, if b is zero.
+     * If a is not zero then by definition of logarithm a^0 is 1.
+     * If a is also zero then as per IEEE 0^0 is best defined as 1.
+     */
     if (br == 0. && bi == 0.) {
         return npy_cpack@c@(1., 0.);
     }
-    if (ar == 0. && ai == 0.) {
-        if (br > 0 && bi == 0) {
-            return npy_cpack@c@(0., 0.);
-        }
-        else {
-            volatile @type@ tmp = NPY_INFINITY@C@;
-            /*
-             * NB: there are four complex zeros; c0 = (+-0, +-0), so that
-             * unlike for reals, c0**p, with `p` negative is in general
-             * ill-defined.
-             *
-             *     c0**z with z complex is also ill-defined.
-             */
-            r = npy_cpack@c@(NPY_NAN@C@, NPY_NAN@C@);
-
-            /* Raise invalid */
-            tmp -= NPY_INFINITY@C@;
-            ar = tmp;
-            return r;
-        }
+    /*
+     * If mantissa is zero, then result depends upon values of 
+     * br and bi. The result is either 0 (in magnitude) or 
+     * undefined ( or 1 for br=bi=0 and independent of ar and ai
+     * but that case is handled above).
+     */
+    else if (ar == 0. && ai == 0.) {
+        /* 
+         * If br > 0 then result is 0 but contains sign opposite
+         * of bi.
+         * Else the result is undefined and we return (nan,nan)
+         */
+         if (br > 0) {
+             if (bi < 0) {
+                 return npy_cpack@c@(0., 0.);
+             }
+             else {
+                 return npy_cpack@c@(0., -0.);
+             }
+             
+         }
+         else {
+             /* 
+              * Raising an invalid and returning
+              * (nan, nan)
+              */
+              volatile @type@ tmp = NPY_INFINITY@C@;
+              r = npy_cpack@c@(NPY_NAN@C@, NPY_NAN@C@);
+
+              /* Raise invalid */
+              tmp -= NPY_INFINITY@C@;
+              ar = tmp;
+              return r;
+         }
```

Comment:
```suggestion
        /* else we are in the case where the
         * real part of b is negative (br<0).
         * Here we should return a complex nan
         * and raise FloatingPointError: invalid value...
         */
         
         /* Raise invalid value by calling inf - inf*/
          volatile @type@ tmp = NPY_INFINITY@C@;
          tmp -= NPY_INFINITY@C@;
          ar = tmp;
          
          r = npy_cpack@c@(NPY_NAN@C@, NPY_NAN@C@);
          return r;
```

Reply:
I am not familiarized with the internals that much. I have some questions

- why do we need `inf` to be `volatile`?
- why do we need to save the `inf-inf` value to `ar`?

Reply:
@prithvitewatia do you have an answer?

Reply:
The `inf - inf` and the volatile is reasonable, although I am not sure it is the best way to spell it.  Maybe we should just use `npy_set_floatstatus_invalid()`, since that is clearer.  The volatile would prevent the compiler from doing the calculation at compile time, which would remove the floating pointer error flags.

The reason that for me this is hard to pick up, is that I think, the branch-cuts were right in the last version with `-0` or not, but I have never seen an IEEE or C standard for *complex* (rather than real) numbers.  So, we have to make a call here and it would be nice to have a second reference somewhere to make that call.